### PR TITLE
Allow annotation functions to handle zero symbol

### PR DESCRIPTION
**Pull request by byronlin-bq**
*Issue #, if available:*
https://github.com/amzn/ion-hash-dotnet/issues/9

*Description of changes:*
Currently the IonHashTests are failing for scenario: $0::{}

This is because IonHashTests calls on the reader.GetTypeAnnotations/HasAnnotations and zero symbol is currently not supported.

Added zero symbol support to the annotation methods

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Amazon.IonDotnet.Tests/Common/ReaderTestCommon.cs [hunk 1]
```diff
@@ -168,11 +168,25 @@ public static void ReadTypeAnnotations_SingleField(IIonReader reader)
...
+        public static void ReadTypeAnnotations_ZeroSymbol(IIonReader reader)
+        {
+            // an int with zero symbol annotation
+            // $0::18
+            reader.MoveNext();
+
+            Assert.AreEqual(IonType.Int, reader.CurrentType);
+            Assert.AreEqual(18, reader.IntValue());
+
+            string[] annotations = reader.GetTypeAnnotations();
```
Comment (tgregg) [inline Amazon.IonDotnet.Tests/Common/ReaderTestCommon.cs]: This should throw `UnknownSymbolException`. No symbols with unknown text, including symbol zero, are accessible via APIs that return strings.

File: Amazon.IonDotnet.Tests/Common/ReaderTestCommon.cs [hunk 2]
```diff
@@ -202,6 +216,20 @@ public static void ReadTypeAnnotationSymbols_SingleField(IIonReader reader)
...
+            // an int with zero symbol annotation
+            // $0::18
+            reader.MoveNext();
+
+            Assert.AreEqual(IonType.Int, reader.CurrentType);
+            Assert.AreEqual(18, reader.IntValue());
+
+            Assert.AreEqual(1, reader.GetTypeAnnotationSymbols().Count());
+
+            Assert.IsTrue(reader.GetTypeAnnotationSymbols().Any(a => a.Text == null));
```
Comment (tgregg) [inline Amazon.IonDotnet.Tests/Common/ReaderTestCommon.cs]: Also need to check that the import location is null.

File: Amazon.IonDotnet.Tests/Common/ReaderTestCommon.cs [hunk 3]
```diff
@@ -250,6 +278,14 @@ public static void HasAnnotationFalse_SingleField(IIonReader reader)
             Assert.IsFalse(reader.HasAnnotation("Spam Musubi"));
         }
 
+        public static void HasAnnotationTrue_ZeroSymbol(IIonReader reader)
+        {
+            // an int with zero symbol annotation
+            // $0::18
+            reader.MoveNext();
+            Assert.IsTrue(reader.HasAnnotation(null));
```
Comment (tgregg) [inline Amazon.IonDotnet.Tests/Common/ReaderTestCommon.cs]: `reader.HasAnnotation(null)` should always be false (or perhaps raise an exception). `null` is not synonymous with symbol zero.

File: Amazon.IonDotnet.Tests/TestDat/text/annot_zerosymbol.ion [hunk 4]
```diff
@@ -0,0 +1 @@
+﻿$0::18
```
Comment (tgregg) [inline Amazon.IonDotnet.Tests/TestDat/text/annot_zerosymbol.ion]: There's really no need to have external files for these. I recommend just defining them inline in the tests. That's much easier to maintain, and makes the tests easier to read.

File: Amazon.IonDotnet.Tests/Common/ReaderTestCommon.cs (hunk 1 above)
Comment (byronlin-bq) [reply Amazon.IonDotnet.Tests/Common/ReaderTestCommon.cs]: Updated assert UnknownSymbolException

File: Amazon.IonDotnet.Tests/Common/ReaderTestCommon.cs (hunk 2 above)
Comment (byronlin-bq) [reply Amazon.IonDotnet.Tests/Common/ReaderTestCommon.cs]: Did you mean SID? When I assert `a.ImportLocation == null`, it fails 

The import location is: `{ImportLocation::{importName:, id:0}}`

File: Amazon.IonDotnet.Tests/Common/ReaderTestCommon.cs (hunk 3 above)
Comment (byronlin-bq) [reply Amazon.IonDotnet.Tests/Common/ReaderTestCommon.cs]: Revert to original, which was raising exception

File: Amazon.IonDotnet.Tests/TestDat/text/annot_zerosymbol.ion (hunk 4 above)
Comment (byronlin-bq) [reply Amazon.IonDotnet.Tests/TestDat/text/annot_zerosymbol.ion]: Removed files

File: Amazon.IonDotnet.Tests/Common/ReaderTestCommon.cs (hunk 2 above)
Comment (tgregg) [reply Amazon.IonDotnet.Tests/Common/ReaderTestCommon.cs]: Looks like a bug. SID 0 should not have an import location.

The fix can be in a separate CR if you want. Please create an issue.

File: Amazon.IonDotnet.Tests/Internals/TreeReaderTest.cs [hunk 5]
```diff
@@ -244,6 +244,17 @@ public void ValueWithTypeAnnotationsTest()
             ReaderTestCommon.ReadTypeAnnotations_SingleField(reader);
         }
 
+        [TestMethod]
+        public void ValueWithTypeAnnotationsTest_ZeroSymbol()
+        {
+            //Must be: $0::18
+            var intValue = _ionValueFactory.NewInt(18);
+            intValue.AddTypeAnnotation(null);
```
Comment (tgregg) [inline Amazon.IonDotnet.Tests/Internals/TreeReaderTest.cs]: `AddTypeAnnotation(null)` should throw an exception. To add symbol zero as an annotation, the user must provide a `SymbolToken`.

File: Amazon.IonDotnet.Tests/Common/ReaderTestCommon.cs (hunk 2 above)
Comment (byronlin-bq) [reply Amazon.IonDotnet.Tests/Common/ReaderTestCommon.cs]: Created issue: https://github.com/amzn/ion-dotnet/issues/89

Could you set milestone?

File: Amazon.IonDotnet.Tests/Internals/TreeReaderTest.cs (hunk 5 above)
Comment (byronlin-bq) [reply Amazon.IonDotnet.Tests/Internals/TreeReaderTest.cs]: Updated to AddTypeAnnotation(SymbolToken)

Comment (byronlin-bq) [reply Amazon.IonDotnet.Tests/Internals/TreeReaderTest.cs]: Also added null check in AddTypeAnnotation(String) and added unit test

**Pull request merged**